File: packages/elixir/lib/doorman/reputation.ex
defmodule Doorman.Reputation do
  @moduledoc "Optional AbuseIPDB check. Short-lived database cache and quota; no IP persistence or reporting."
  alias Doorman.Storage, as: S
  def configure(nil, _), do: nil

  def configure(opts, identity) do
    unless identity && Keyword.keyword?(opts) &&
             Enum.all?(
               Keyword.keys(opts),
               &(&1 in [
                   :api_key,
                   :timeout_ms,
                   :cache_ttl_ms,
                   :max_requests_per_hour,
                   :req_options
                 ])
             ),
           do: raise(ArgumentError, "invalid reputation configuration")

    unless is_binary(opts[:api_key]) && String.trim(opts[:api_key]) != "",
      do: raise(ArgumentError, "reputation API key is required")

    opts =
      Keyword.merge([timeout_ms: 500, cache_ttl_ms: 300_000, max_requests_per_hour: 100], opts)

    for {key, min, max} <- [
          {:timeout_ms, 50, 3000},
          {:cache_ttl_ms, 1000, 3_600_000},
          {:max_requests_per_hour, 1, 10_000}
        ] do
      unless is_integer(opts[key]) && opts[key] >= min && opts[key] <= max,
        do: raise(ArgumentError, "invalid reputation limit")
    end

    # Req.Test hooks only; never accept a configurable request destination.
    unless Enum.all?(Keyword.keys(opts[:req_options] || []), &(&1 in [:plug])),
      do: raise(ArgumentError, "invalid reputation transport options")

    opts
  end

  def check(%{reputation: nil}, _), do: nil

  def check(c, raw) do
    case public_ip(raw) do
      nil ->
        empty("not-requested")

      ip ->
        key = label(c, ["reputation-cache-v1", ip])
        now = Doorman.now()

        case S.query(
               c,
               "SELECT record FROM #{S.table(c, "evaluation_controls")} WHERE id=$1 AND expires_at > $2",
               [key, now]
             ) do
          [%{"record" => value}] -> Map.put(value, "cached", true)
          [] -> lookup(c, ip, key, now)
        end
    end
  rescue
    _ -> empty("unavailable")
  catch
    _, _ -> empty("unavailable")
  end

  defp lookup(c, ip, key, now) do
    opts = c.reputation

    lease =
      S.query(
        c,
        "INSERT INTO #{S.table(c, "evaluation_controls")} AS e (id,version,expires_at,record) VALUES ($1,$2,$3,$4) ON CONFLICT (id) DO UPDATE SET version=EXCLUDED.version,expires_at=EXCLUDED.expires_at,record=EXCLUDED.record WHERE e.expires_at <= $2 RETURNING id",
        [key, now, now + opts[:timeout_ms] + 1000, empty("unavailable")]
      )

    if lease == [] do
      empty("unavailable")
    else
      quota =
        S.query(
          c,
          "INSERT INTO #{S.table(c, "protection_quotas")} AS q (id,used,reset_at) VALUES ($1,1,$2) ON CONFLICT (id) DO UPDATE SET used=CASE WHEN q.reset_at <= $3 THEN 1 ELSE q.used+1 END,reset_at=CASE WHEN q.reset_at <= $3 THEN $2 ELSE q.reset_at END WHERE q.reset_at <= $3 OR q.used < $4 RETURNING id",
          [label(c, ["reputation-budget-v1"]), now + 3_600_000, now, opts[:max_requests_per_hour]]
        )

      value =
        if quota == [] do
          empty("limited")
        else
          case Doorman.Bounded.run(fn -> fetch(ip, opts) end, opts[:timeout_ms]) do
            {:ok, value} -> value
            _ -> empty("unavailable")
          end
        end

      ttl = if value["status"] == "available", do: opts[:cache_ttl_ms], else: 30_000

      S.query(
        c,
        "UPDATE #{S.table(c, "evaluation_controls")} SET record=$1,expires_at=$2 WHERE id=$3 AND version=$4",
        [value, Doorman.now() + ttl, key, now]
      )

      value
    end
  end

  defp fetch(ip, opts) do
    response =
      Doorman.HTTP.get_json(
        Keyword.merge(opts[:req_options] || [],
          url: "https://api.abuseipdb.com/api/v2/check",
          params: [ipAddress: ip, maxAgeInDays: 30],
          headers: [{"key", opts[:api_key]}, {"accept", "application/json"}],
          retry: false,
          redirect: false,
          receive_timeout: opts[:timeout_ms],
          connect_options: [timeout: opts[:timeout_ms]]
        )
      )

    data = response.body["data"]

    unless response.status == 200 && is_map(data) && public_ip(data["ipAddress"]) == ip &&
             data["isPublic"] == true && is_integer(data["abuseConfidenceScore"]) &&
             data["abuseConfidenceScore"] in 0..100 && is_integer(data["totalReports"]) &&
             data["totalReports"] in 0..1_000_000_000,
           do: raise("invalid reputation response")

    value =
      Map.merge(empty("available"), %{
        "score" => data["abuseConfidenceScore"] / 100,
        "totalReports" => data["totalReports"]
      })

    case DateTime.from_iso8601(data["lastReportedAt"] || "") do
      {:ok, time, _} ->
        ms = DateTime.to_unix(time, :millisecond)

        if ms >= 0 && ms <= Doorman.now() + 5000,
          do: Map.put(value, "lastReportedAt", ms),
          else: value

      _ ->
        value
    end
  end

  defp public_ip(raw) when is_binary(raw) do
    case :inet.parse_address(String.to_charlist(raw)) do
      {:ok, {a, b, _, _} = address}
      when a != 0 and a != 10 and a != 127 and a < 224 and not (a == 169 and b == 254) and
             not (a == 172 and b >= 16 and b <= 31) and not (a == 192 and b == 168) and
             not (a == 100 and b >= 64 and b <= 127) and not (a == 198 and b in [18, 19]) ->
        to_string(:inet.ntoa(address))

      {:ok, {a, _, _, _, _, _, _, _} = address} when a >= 0x2000 and a <= 0x3FFF ->
        to_string(:inet.ntoa(address))

      _ ->
        nil
    end
  end

  defp public_ip(_), do: nil
  defp label(c, parts), do: Doorman.Identity.label(c.identity, Jason.encode!(parts))

  defp empty(status),
    do: %{
      "provider" => "abuseipdb",
      "status" => status,
      "cached" => false,
      "observedAt" => Doorman.now()
    }
end

File: packages/adapters/src/reputation.ts
import { z } from "zod";
import type {
  ReputationEvidence,
  RiskEvidence,
} from "@aarondovturkel/doorman-core";
import { createSubjectLinker, type SubjectLinkingOptions } from "./subject.js";

export type ReputationOptions = {
  /** Enables sharing the server-resolved client IP with AbuseIPDB. */
  apiKey: string;
  timeoutMs?: number;
  cacheTtlMs?: number;
  /** Shared database budget in server adapters; also set the provider account quota. */
  maxRequestsPerHour?: number;
};
const probability = z.number().finite().min(0).max(1);
export const riskEvidenceSchema = z.strictObject({
  edge: z
    .strictObject({
      source: z.literal("edge"),
      provider: z.string().max(64),
      observedAt: z.number().int().nonnegative(),
      botScore: z.number().int().min(1).max(99).optional(),
      verifiedBot: z.boolean().optional(),
      signedAgent: z.boolean().optional(),
      ja4: z
        .string()
        .regex(/^[tqd][a-z0-9]{9}_[a-f0-9]{12}_[a-f0-9]{12}$/)
        .optional(),
    })
    .optional(),
  activity: z
    .strictObject({
      windowMs: z.number().int().min(1000).max(86_400_000),
      requests: z.number().int().min(0).max(1_000_000),
      denials: z.number().int().min(0).max(1_000_000),
      authenticationFailures: z.number().int().min(0).max(1_000_000),
    })
    .refine(
      (a) => a.denials <= a.requests && a.authenticationFailures <= a.requests,
    )
    .optional(),
  reputation: z
    .strictObject({
      provider: z.literal("abuseipdb"),
      status: z.enum(["available", "unavailable", "limited", "not-requested"]),
      observedAt: z.number().int().nonnegative(),
      cached: z.boolean(),
      score: probability.optional(),
      totalReports: z
        .number()
        .int()
        .nonnegative()
        .max(1_000_000_000)
        .optional(),
      lastReportedAt: z.number().int().nonnegative().optional(),
    })
    .optional(),
});
export function trustedRiskEvidence(
  input?: RiskEvidence,
): RiskEvidence | undefined {
  if (!input) return undefined;
  const result = riskEvidenceSchema.parse(input);
  const now = Date.now();
  if (
    result.edge &&
    (result.edge.observedAt > now + 5000 ||
      result.edge.observedAt < now - 300_000)
  )
    delete result.edge;
  if (
    result.reputation &&
    (result.reputation.observedAt > now + 5000 ||
      result.reputation.observedAt < now - 3_600_000)
  )
    delete result.reputation;
  return result;
}

function publicIp(value: unknown): string | undefined {
  if (typeof value !== "string") return undefined;
  if (z.ipv4().safeParse(value).success) {
    const [a, b] = value.split(".").map(Number) as [number, number];
    if (
      a === 0 ||
      a === 10 ||
      a === 127 ||
      a >= 224 ||
      (a === 169 && b === 254) ||
      (a === 172 && b >= 16 && b <= 31) ||
      (a === 192 && b === 168) ||
      (a === 100 && b >= 64 && b <= 127) ||
      (a === 198 && (b === 18 || b === 19))
    )
      return undefined;
    return value;
  }
  if (z.ipv6().safeParse(value).success && /^[23][a-f0-9]{0,3}:/i.test(value))
    return new URL(`https://[${value}]/`).hostname.slice(1, -1);
  return undefined;
}

/** Read-only reputation enrichment. Raw IPs never enter the cache, result, database or Jev. */
export function createReputation(
  options: ReputationOptions,
  identity: SubjectLinkingOptions,
  admit?: () => Promise<boolean>,
) {
  const config = z
    .strictObject({
      apiKey: z.string().trim().min(1).max(4096),
      timeoutMs: z.number().int().min(50).max(3000).default(500),
      cacheTtlMs: z.number().int().min(1000).max(3_600_000).default(300_000),
      maxRequestsPerHour: z.number().int().min(1).max(10_000).default(100),
    })
    .parse(options);
  const label = createSubjectLinker(identity);
  const cache = new Map<
    string,
    { expires: number; value: ReputationEvidence }
  >();
  const pending = new Map<string, Promise<ReputationEvidence>>();
  let period = 0,
    calls = 0;
  const empty = (status: ReputationEvidence["status"]): ReputationEvidence => ({
    provider: "abuseipdb",
    status,
    observedAt: Date.now(),
    cached: false,
  });
  return {
    async check(value?: string): Promise<ReputationEvidence> {
      const ip = publicIp(value);
      if (!ip) return empty("not-requested");
      const key = await label(JSON.stringify(["reputation-v1", ip]));
      const found = cache.get(key);
      if (found && found.expires > Date.now())
        return { ...found.value, cached: true };
      if (pending.has(key))
        return { ...(await pending.get(key)!), cached: true };
      const hour = Math.floor(Date.now() / 3_600_000);
      if (hour !== period) {
        period = hour;
        calls = 0;
      }
      if (calls >= config.maxRequestsPerHour || pending.size >= 8)
        return empty("limited");
      calls++;
      const work = (async () => {
        let result = empty("unavailable");
        try {
          if (admit && !(await admit())) return empty("limited");
          const url = new URL("https://api.abuseipdb.com/api/v2/check");
          url.searchParams.set("ipAddress", ip);
          url.searchParams.set("maxAgeInDays", "30");
          const response = await fetch(url, {
            headers: { Key: config.apiKey, Accept: "application/json" },
            signal: AbortSignal.timeout(config.timeoutMs),
            redirect: "error",
          });
          if (!response.ok) {
            await response.body?.cancel();
            throw new Error("Reputation unavailable");
          }
          const reader = response.body?.getReader();
          if (!reader) throw new Error("Empty reputation response");
          const decoder = new TextDecoder();
          let text = "",
            size = 0;
          try {
            for (;;) {
              const { done, value } = await reader.read();
              if (done) break;
              size += value.byteLength;
              if (size > 8192) throw new Error("Reputation response too large");
              text += decoder.decode(value, { stream: true });
            }
          } finally {
            await reader.cancel().catch(() => {});
          }
          const raw: unknown = JSON.parse(text + decoder.decode());
          const data = z
            .object({
              data: z.object({
                ipAddress: z.string(),
                isPublic: z.literal(true),
                abuseConfidenceScore: z.number().int().min(0).max(100),
                totalReports: z.number().int().min(0).max(1_000_000_000),
                lastReportedAt: z.string().max(64).nullable().optional(),
              }),
            })
            .parse(raw).data;
          if (publicIp(data.ipAddress) !== ip)
            throw new Error("Reputation address mismatch");
          const last = data.lastReportedAt
            ? Date.parse(data.lastReportedAt)
            : undefined;
          result = {
            ...empty("available"),
            score: data.abuseConfidenceScore / 100,
            totalReports: data.totalReports,
            ...(last !== undefined &&
            Number.isFinite(last) &&
            last >= 0 &&
            last <= Date.now() + 5000
              ? { lastReportedAt: last }
              : {}),
          };
        } catch {
          /* Optional enrichment never interrupts identity. */
        }
        if (cache.size >= 1000) cache.delete(cache.keys().next().value!);
        cache.set(key, {
          expires:
            Date.now() +
            (result.status === "available" ? config.cacheTtlMs : 30_000),
          value: result,
        });
        return result;
      })();
      pending.set(key, work);
      try {
        return await work;
      } finally {
        pending.delete(key);
      }
    },
  };
}
